discord_connector.py

```diff
@@ -309,6 +383,86 @@ def run_in_loop(self, coroutine, action_result, error_message=""):
+        self._status = self.load_state()
+        last_poll_date = self._state.get("last_poll_date", None)
+        if last_poll_date is not None:
+            last_poll_date_parsing_status, last_poll_date = self.parse_date(last_poll_date)
+            if not last_poll_date_parsing_status:
+                return action_result.set_status(phantom.APP_ERROR, "Unable to parse last poll date")
+
+        self.save_progress("lading last poll date: {}".format(last_poll_date))
+
+        newest_message = last_poll_date
+        if newest_message is None:
+            newest_message = datetime.min.replace(tzinfo=pytz.UTC)
+
+        for channel in filter((lambda channel_to_test: isinstance(channel_to_test, discord.TextChannel)), channels):
+
+            status, messages = self.fetch_message_history(channel.id, action_result, last_poll_date,
+                                                          None, True, container_count)
+            if not status:
+                self.debug_print("action result: Cannot Poll messages from {} {}".format(channel.name, channel.id))
+                return action_result.set_status(phantom.APP_ERROR)
+
+            self.save_progress("saving the containers")
+            for message in messages:
+
+                _, message_creation_date = self.parse_date(message.created_at.strftime("%Y-%m-%d %H:%M:%S"))
+                if newest_message < message_creation_date:
+                    newest_message = message_creation_date
+
+                if message_creation_date != last_poll_date:
+                    ret_val = self.save_on_poll_container(message, channel)
+                    if phantom.is_fail(ret_val):
+                        return action_result.set_status(phantom.APP_ERROR, "Unable to create container: {}".format(message))
+
+        newest_message = newest_message.replace(tzinfo=pytz.UTC)
+        self.save_progress("saving last poll date: {}".format(str(newest_message)[:-6]))
+        self._state["last_poll_date"] = str(newest_message)[:-6]
+        self.save_state(self._state)
+
+        action_result = self.add_action_result(ActionResult(dict(param)))
+        return action_result.set_status(phantom.APP_SUCCESS)
+
+    def save_on_poll_container(self, message, channel):
+
+        contains_embeds_or_attachments = True if (message.attachments or message.embeds) else False
+
+        container = {
+            "name": f"message {message.id} on channel {channel.name}",
+            "description": "generated by on poll action",
+            "sensitivity": "White" if not contains_embeds_or_attachments else "Green"
+        }
+
+        ret_val, ret_message, container_id = self.save_container(container)
+
+        self.save_progress("container saved finished: return value: {} | message: {} | container_id: {}"
+                           .format(ret_val, ret_message, container_id))
+
+        artifact = Artifact(
+            container_id=str(container_id),
+            name=str(message.id),
+            cef={
```

Comment:
why not use `Cef` dataclass?

Reply:
OK, so this is a different set of fields in this CEF, I believe?

Reply:
Yes, those are different. I can create dataclass for that too, but I believe that it would argue with object oriented programming if we would use it only once.